spec/features/sign_in_spec.rb

```diff
@@ -55,26 +63,67 @@
       end
     end
 
-    context 'with username' do
+    context 'with uid' do
       before do
-        fill_in 'Username or email', with: user.username
-        fill_in 'Password', with: user.password
-        click_button 'Sign in'
+        user.update(entra_id: '19846')
+        OmniAuth.config.mock_auth[:entra] = OmniAuth::AuthHash.new({
+                                                                     uid: '19846',
+                                                                     info: {
+                                                                       'email' => 'bob.smith.new@example.com'
+                                                                     }
+                                                                   })
+        click_button 'Sign in with your Ministry of Justice account'
       end
 
       it 'successful sign in message displayed' do
         expect(page).to have_govuk_flash(:notice, text: 'Signed in successfully')
       end
+
+      it 'updates email' do
+        expect(user.reload.email).to eq 'bob.smith.new@example.com'
+      end
     end
   end
 
-  context 'with failure' do
+  context 'when user has no matching account' do
+    before do
+      OmniAuth.config.mock_auth[:entra] = OmniAuth::AuthHash.new({
+                                                                   uid: '19846',
+                                                                   info: {
+                                                                     'email' => 'bob.smith.new@example.com'
+                                                                   }
+                                                                 })
+      click_button 'Sign in with your Ministry of Justice account'
+    end
+
     it 'invalid username, email or password displayed' do
-      fill_in 'Username or email', with: 'billy bob'
-      fill_in 'Password', with: user.password
-      click_button 'Sign in'
       expect(page).to have_govuk_page_heading(text: 'Sign in')
-      expect(page).to have_govuk_flash(:alert, text: 'Invalid username or password')
+      expect(page).to have_govuk_flash(:alert, text: 'You do not have permission to access this service')
+    end
+  end
+
+  context 'with an unsuccessful omniauth flow' do
+    before do
+      OmniAuth.config.mock_auth[:entra] = :invalid_credentials
+      click_button 'Sign in with your Ministry of Justice account'
+    end
+
+    it 'invalid username, email or password displayed' do
```

Comment:
Active tense instead of passive?
`it 'displays invalid username, email or password'`